gems/smithy-client/lib/smithy-client/net_http/connection_pool.rb

```diff
@@ -0,0 +1,279 @@
+          logger: Logger.new($stdout)
+        }.freeze
+
+        # @api private
+        def initialize(options = {})
+          OPTIONS.each_pair do |opt_name, default_value|
+            value = options[opt_name].nil? ? default_value : options[opt_name]
+            instance_variable_set("@#{opt_name}", value)
+          end
+          @pool_mutex = Mutex.new
+          @pool = {}
+        end
+        private_class_method :new
+
+        OPTIONS.each_key do |attr_name|
+          attr_reader(attr_name)
+        end
+
+        # @param [URI::HTTP, URI::HTTPS] endpoint The HTTP(S) endpoint
+        #    to connect to (e.g. 'https://domain.com').
+        #
+        # @yieldparam [Net::HTTPSession] session
+        #
+        # @return [nil]
+        def session_for(endpoint)
+          endpoint = remove_path_and_query(endpoint)
+          session = nil
+
+          # attempt to recycle an already open session
+          @pool_mutex.synchronize do
+            _clean
+            session = @pool[endpoint].shift if @pool.key?(endpoint)
+          end
+
+          begin
+            session ||= start_session(endpoint)
+            session.read_timeout = http_read_timeout if http_read_timeout
+            session.continue_timeout = http_continue_timeout if http_continue_timeout
+            yield(session)
+          rescue StandardError
+            session&.finish
+            raise
+          else
+            @pool_mutex.synchronize do
+              @pool[endpoint] = [] unless @pool.key?(endpoint)
+              @pool[endpoint] << session
+            end
+          end
+          nil
+        end
+
+        # @return [Integer] Returns the count of sessions currently in the
+        #  pool, not counting those currently in use.
+        def size
+          @pool_mutex.synchronize do
+            @pool.values.flatten.size
+          end
+        end
+
+        # Removes stale http sessions from the pool (that have exceeded the idle timeout).
+        # @return [nil]
+        def clean!
+          @pool_mutex.synchronize { _clean }
+          nil
+        end
+
+        # Closes and removes all sessions from the pool. If empty! is called while
+        # there are outstanding requests they may get checked back into the pool,
+        # leaving the pool in a non-empty state.
+        # @return [nil]
+        def empty!
+          @pool_mutex.synchronize do
+            @pool.values.flatten.map(&:finish)
+            @pool.clear
+          end
+          nil
+        end
+
+        private
+
+        def remove_path_and_query(endpoint)
+          endpoint.dup.tap do |e|
+            e.path = ''
+            e.query = nil
+          end.to_s
+        end
+
+        class << self
+          # Returns a connection pool constructed from the given options.
+          # Calling this method twice with the same options will return
+          # the same pool.
+          #
+          # @option options [Numeric] :http_continue_timeout (nil)
```

Comment:
we probably don't really need all of this documentation here since this is api private? should probably just live in the http.... handler? We don't really have a client class right?

Edit: NVM, I guess maybe this is the best place for this documentation since this class is the one actually using them.  I'm not sure, I find it a little weird to have documented here.

Reply:
It's mostly for us.